File: android/app/src/main/kotlin/com/mayasabhaxr/recapture/storage/JobManifest.kt
// android/app/src/main/kotlin/com/mayasabhaxr/recapture/storage/JobManifest.kt
package com.mayasabhaxr.recapture.storage

import com.mayasabhaxr.recapture.camera.MetadataJson

/**
 * Per-job manifest/marker used to distinguish a COMPLETE job from one INTERRUPTED
 * mid-capture (app killed during a burst). It is written at job start
 * ([Status.IN_PROGRESS]) and finalized at completion ([Status.COMPLETE]); a job dir
 * whose manifest is missing or not [Status.COMPLETE] is "incomplete" and can be
 * resumed or cleaned (see [CaptureStorage.listIncompleteJobs]).
 *
 * Encoded with the shared [MetadataJson] encoder for consistency with the per-frame
 * sidecars; decoded by a small targeted reader over the few, controlled scalar keys
 * (the values are sanitized ids/numbers/one of two status tokens — no arbitrary JSON),
 * so both encode and parse are pure and JVM-unit-testable without a JSON runtime.
 */
data class JobManifest(
    val projectId: String,
    val jobId: String,
    val status: Status,
    val startedAtMs: Long,
    val completedAtMs: Long? = null,
    /** Frames the capture flow reported at completion (informational; null until complete). */
    val frameCount: Int? = null,
) {
    enum class Status(val wire: String) {
        IN_PROGRESS("in_progress"),
        COMPLETE("complete");

        companion object {
            fun fromWire(s: String?): Status? = entries.firstOrNull { it.wire == s }
        }
    }

    val isComplete: Boolean get() = status == Status.COMPLETE

    fun encode(): String {
        val map = LinkedHashMap<String, Any?>()
        map["projectId"] = projectId
        map["jobId"] = jobId
        map["status"] = status.wire
        map["startedAtMs"] = startedAtMs
        map["completedAtMs"] = completedAtMs
        map["frameCount"] = frameCount
        return MetadataJson.encode(map)
    }

    companion object {
        /** Manifest file name; lives in the JOB dir (above the `images/` levels). */
        const val FILE_NAME = "_manifest.json"

        /**
         * Parses a manifest written by [encode]. Tolerant of key order and absent
         * optional fields; returns null only if the required identity/status/start
         * fields are unreadable (then the job is treated as incomplete, never trusted).
         */
        fun parse(text: String): JobManifest? {
            val projectId = stringField(text, "projectId") ?: return null
            val jobId = stringField(text, "jobId") ?: return null
            val status = Status.fromWire(stringField(text, "status")) ?: return null
            val startedAtMs = longField(text, "startedAtMs") ?: return null
            return JobManifest(
                projectId = projectId,
                jobId = jobId,
                status = status,
                startedAtMs = startedAtMs,
                completedAtMs = longField(text, "completedAtMs"),
                frameCount = longField(text, "frameCount")?.toInt(),
            )
        }

        private fun stringField(text: String, key: String): String? =
            Regex("\"$key\"\\s*:\\s*\"([^\"]*)\"").find(text)?.groupValues?.get(1)

        // Number field; explicit `null` (or absent) → null.
        private fun longField(text: String, key: String): Long? =
            Regex("\"$key\"\\s*:\\s*(-?\\d+)").find(text)?.groupValues?.get(1)?.toLongOrNull()
    }
}

File: ios/Runner/JobManifest.swift
import Foundation

/// Per-job manifest/marker used to distinguish a COMPLETE job from one INTERRUPTED
/// mid-capture (app killed during a burst) — the iOS counterpart to the Android
/// `JobManifest`, kept wire-compatible (same keys, same `in_progress`/`complete`
/// tokens, same `_manifest.json` name) so either platform can read the other's tree
/// and `CaptureStorage.listIncompleteJobs` behaves identically.
///
/// Written at job start (`inProgress`) and finalized at completion (`complete`); a
/// job dir whose manifest is missing or not `complete` is "incomplete" and can be
/// resumed or cleaned. Encode produces a small JSON object over controlled scalar
/// keys (sanitized ids / numbers / one of two status tokens — no arbitrary JSON), so
/// both encode and parse are pure and unit-testable without a JSON runtime.
struct JobManifest {

  enum Status: String {
    case inProgress = "in_progress"
    case complete = "complete"

    static func fromWire(_ s: String?) -> Status? {
      guard let s = s else { return nil }
      return Status(rawValue: s)
    }
  }

  let projectId: String
  let jobId: String
  let status: Status
  let startedAtMs: Int64
  /// Set only once the job completes.
  let completedAtMs: Int64?
  /// Frames the capture flow reported at completion (informational; nil until complete).
  let frameCount: Int?

  init(
    projectId: String, jobId: String, status: Status, startedAtMs: Int64,
    completedAtMs: Int64? = nil, frameCount: Int? = nil
  ) {
    self.projectId = projectId
    self.jobId = jobId
    self.status = status
    self.startedAtMs = startedAtMs
    self.completedAtMs = completedAtMs
    self.frameCount = frameCount
  }

  var isComplete: Bool { status == .complete }

  /// Manifest file name; lives in the JOB dir (above the `images/` levels).
  static let fileName = "_manifest.json"

  /// Encodes to the JSON shape `parse` reads. `projectId`/`jobId` are allowlisted
  /// (no quotes/backslashes), so manual assembly is safe and dependency-free.
  func encode() -> String {
    let completed = completedAtMs.map { String($0) } ?? "null"
    let frames = frameCount.map { String($0) } ?? "null"
    return "{"
      + "\"projectId\":\"\(projectId)\","
      + "\"jobId\":\"\(jobId)\","
      + "\"status\":\"\(status.rawValue)\","
      + "\"startedAtMs\":\(startedAtMs),"
      + "\"completedAtMs\":\(completed),"
      + "\"frameCount\":\(frames)"
      + "}"
  }

  /// Parses a manifest written by `encode`. Tolerant of key order and absent
  /// optional fields; returns nil only if the required identity/status/start fields
  /// are unreadable (then the job is treated as incomplete, never trusted).
  static func parse(_ text: String) -> JobManifest? {
    guard let projectId = stringField(text, "projectId"),
          let jobId = stringField(text, "jobId"),
          let status = Status.fromWire(stringField(text, "status")),
          let startedAtMs = longField(text, "startedAtMs") else {
      return nil
    }
    let completed = longField(text, "completedAtMs")
    let frames = longField(text, "frameCount").map { Int($0) }
    return JobManifest(
      projectId: projectId, jobId: jobId, status: status, startedAtMs: startedAtMs,
      completedAtMs: completed, frameCount: frames)
  }

  private static func stringField(_ text: String, _ key: String) -> String? {
    return firstGroup(in: text, pattern: "\"\(key)\"\\s*:\\s*\"([^\"]*)\"")
  }

  // Number field; explicit `null` (or absent) → nil.
  private static func longField(_ text: String, _ key: String) -> Int64? {
    return firstGroup(in: text, pattern: "\"\(key)\"\\s*:\\s*(-?\\d+)").flatMap { Int64($0) }
  }

  private static func firstGroup(in text: String, pattern: String) -> String? {
    guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
    let range = NSRange(text.startIndex..., in: text)
    guard let m = re.firstMatch(in: text, range: range), m.numberOfRanges >= 2,
          let r = Range(m.range(at: 1), in: text) else { return nil }
    return String(text[r])
  }
}
